src/merge.c

```diff
@@ -3239,6 +3239,10 @@ int git_merge(
 
 	assert(repo && their_heads);
 
+	if((error = git_repository_index(&index, repo) < 0) ||
```

Comment:
This actually needs to move down in the code a bit, write _after_ `git_indexwriter_init_for_operation`.

`git_indexwriter` will take set up the lockfile on the index.  We should re-read the index under that lock - if we do not, there's still a race condition where we read the index, and somebody makes changes before we take the lock.